rfcs/proposed/flow_graph_fold_node/README.md

```diff
@@ -0,0 +1,226 @@
+
+```cpp
+tbb::flow::graph g;
+
+tbb::flow::fold_node<int, int> f_node(g, tbb::flow::unlimited,
+    /*fold initializer = */0,
+    /*fold operation = */std::plus<int>{});
+```
+
+Since the ``fold_node`` needs to operate multiple _fold input stream_s in parallel, some identifier of the stream should be provided
+as an input message together with the actual data to compute. It is proposed to use ``tbb::flow::tagged_msg`` class for that purpose
+as an input for the ``fold_node`` that would be sent by the predecessor or by manual ``try_put`` calls:
+
+```cpp
+tbb::flow::graph g;
+
+tbb::flow::fold_node<int, int> f_node(g, tbb::flow::unlimited,
+    /*fold initializer = */0,
+    /*fold operation = */std::plus<int>{});
+
+tbb::flow::buffer_node<int> buf(g);
+
+tbb::flow::make_edge(f_node, buf);
+
+std::size_t stream1_index = 1, stream2_index = 2;
+
+std::vector<int> stream1 = {1, 3, 5, 7, 9};
+std::vector<int> stream2 = {2, 4, 6, 8, 10};
+
+using fold_input = typename tbb::flow::fold_node<int, int>::input_type;
+
+// Submit stream1
+for (auto item : stream1) {
+    f_node.try_put(input_type{stream1_index, item});
+}
+
+for (auto item : stream2) {
+    f_node.try_put(input_type{stream2_index, item});
+}
+
+int result = 0;
+buf.try_get(result);
+
+// result is expected to be either 25 (result of the first stream fold) or
+// 30 (result of the second stream fold)
+```
+
+Explicit specification the ``tagged_msg`` as an ``Input`` template parameter as well as allowing ``tagged_msg`` as a fold operation argument
+are [open question](#process-specific-information).
+
+An other important aspect is since the ``fold_node`` takes the elements from the _fold input stream_ one-by-one, not the entire stream and
+computes partial folds for each element, it is unclear for the node when the computed result is final and all of the elements in the stream
+were processed. To handle this, it is required to take a special signal from the user indicating that no more elements from the input stream
+with the specified tag are expected. Some special type defined by the Flow Graph is required for that purpose, e.g. ``tbb::flow::fold_stream_end``.
+``fold_stream_end`` instance should be explicitly submitted to the node once the input stream end was reached:
+
+```cpp
+// Submit stream1
+for (auto item : stream1) {
+    f_node.try_put(input_type{stream1_index, item});
+}
+f_node.try_put(tbb::flow::fold_stream_end{stream1_index});
+
+for (auto item : stream2) {
+    f_node.try_put(input_type{stream2_index, item});
+}
+f_node.try_put(tbb::flow::fold_stream_end{stream2_index});
+
+int result = 0;
+buf.try_get(result);
+
+// result is expected to be either 25 (result of the first stream fold) or
+// 30 (result of the second stream fold)
+```
+
+From the implementation perspective, there can be multiple approaches for implementing the ``fold_stream_end`` that affects the user API.
+Some of them were considered in a [separate section](#fold_stream_end-implementation-approaches).
```

Comment:
It seems that all the approaches are based on a message counter. The question then is, at which point the total number of messages to be folded (the stream "size") becomes known. If it is known a priori, or if we want to compute partial folds for each K messages, the counter might be set even at construction. If the stream size is dynamic but known before the first message, we may require that the stream is somehow "initialized" to set that counter.

The most flexible, but also the most complicated and risky approach is to get the stream size at any point, or to get the end of stream message. If the message does not carry any counter, there are problems with potential "still in the flight" messages that might be missed. If the message does carry a counter, what to do if more messages were already received and folded? Also, the message type is kind of problematic, as both the stream size and the end of stream message are likely of different type than the stream data.

A possible design alternative is to have a dedicated port for the "end of stream" signals.